File: tools/teleport_rapid.py
# -*- coding: utf-8 -*-
"""Rapid teleport: find addr -> auto teleport test"""
import sys, time, subprocess, os

ADB = r"D:\Setup_and_Downloads\Setup\MuMuPlayer\nx_main\adb.exe"
SERIAL = "127.0.0.1:16480"

r = subprocess.run([ADB, "-s", SERIAL, "shell", "ps", "-A"],
                   capture_output=True, text=True, timeout=15)
pid = None
for line in r.stdout.split("\n"):
    if "proj.xqj" in line:
        parts = line.split()
        if len(parts) >= 2:
            pid = int(parts[1])
            break
print(f"PID={pid}", flush=True)

game_fd = -1
for tcp_file in ["net/tcp", "net/tcp6"]:
    r = subprocess.run([ADB, "-s", SERIAL, "shell", f"cat /proc/{pid}/{tcp_file}"],
                       capture_output=True, text=True, timeout=10)
    for line in r.stdout.split("\n"):
        line = line.strip()
        if not line or line.startswith("sl"):
            continue
        parts = line.split()
        if len(parts) >= 10 and parts[3] == "01":
            inode = parts[9]
            if inode != "0":
                r2 = subprocess.run([ADB, "-s", SERIAL, "shell", f"ls -la /proc/{pid}/fd/ | grep {inode}"],
                                   capture_output=True, text=True, timeout=10)
                for fl in r2.stdout.split("\n"):
                    fp = fl.strip().split()
                    if len(fp) >= 8:
                        try:
                            fd = int(fp[7])
                            if fd > 2:
                                game_fd = fd
                                break
                        except:
                            pass
        if game_fd > 0:
            break
    if game_fd > 0:
        break
print(f"Game fd={game_fd}", flush=True)
if game_fd < 0:
    sys.exit(1)

subprocess.run([ADB, "-s", SERIAL, "forward", "tcp:27055", "tcp:27042"],
               capture_output=True, timeout=10)

script_dir = os.path.dirname(os.path.abspath(__file__))
with open(os.path.join(script_dir, 'teleport_rapid.js'), 'r') as f:
    js_template = f.read()
JS = js_template % game_fd

import frida
dev = frida.get_device_manager().add_remote_device("127.0.0.1:27055")
session = dev.attach(pid)

last_x, last_y = None, None
found_addr = None
found_event = False
tp_active = False

def on_msg(msg, data):
    global last_x, last_y, found_addr, found_event, tp_active
    payload = msg.get('payload', {})
    if not isinstance(payload, dict):
        if msg.get('type') == 'error':
            print(f"[!] {msg.get('description', msg)}", flush=True)
        return
    ptype = payload.get('t', '?')

    if ptype == 'ready':
        print("[*] 就绪 - 移动两次找到地址，然后自动测试瞬移", flush=True)

    elif ptype == 'scan1':
        print(f"[SCAN1] ({payload['x']},{payload['y']}) -> {payload['count']} matches", flush=True)

    elif ptype == 'scan2':
        print(f"[SCAN2] ({payload['x']},{payload['y']}) -> {payload['count']} matches", flush=True)

    elif ptype == 'found':
        found_addr = payload['addr']
        found_event = True
        print(f"\n[!!!] 找到: {payload['addr']} at ({payload['x']},{payload['y']})", flush=True)

    elif ptype == 'retry':
        print(f"[-] 未找到交叉匹配，继续移动...", flush=True)

    elif ptype == 'pos':
        x, y = payload['x'], payload['y']
        mem = payload.get('mem', '')
        tp = '[TP] ' if payload.get('target') else ''
        if last_x is None or x != last_x or y != last_y:
            print(f"  {tp}pos=({x},{y}){mem}", flush=True)
            last_x, last_y = x, y

    elif ptype == 'tp_start':
        tp_active = True
        print(f"[TP] 持续写入 ({payload['x']},{payload['y']}) @ {payload['addr']} (每30ms)", flush=True)

    elif ptype == 'tp_noaddr':
        print(f"[TP] 无地址，仅改包: ({payload['x']},{payload['y']})", flush=True)

    elif ptype == 'tp_stop':
        tp_active = False
        print(f"[TP] 停止", flush=True)

    elif ptype == 'speed_set':
        print(f"[SPEED] {payload['mult']}x", flush=True)

script = session.create_script(JS)
script.on('message', on_msg)
script.load()

# Phase 1: Wait for address
print("Phase 1: 移动两次找到位置地址...", flush=True)
deadline = time.time() + 60
while not found_event and time.time() < deadline:
    time.sleep(0.3)

if not found_event:
    print("[-] 超时，未找到地址", flush=True)
    session.detach()
    sys.exit(1)

# Phase 2: Wait for stable position, then test teleport
print("\nPhase 2: 等待稳定位置后测试瞬移...", flush=True)
time.sleep(3)

if last_x is None:
    print("[-] 未检测到位置", flush=True)
    session.detach()
    sys.exit(1)

# Test 1: Teleport +50 in X and Y
tx = (last_x + 50) & 0xFF
ty = (last_y + 50) & 0xFF
print(f"\n[TEST1] 瞬移: ({last_x},{last_y}) -> ({tx},{ty}) (持续写入+改包)", flush=True)
old_x, old_y = last_x, last_y
script.post({'type': 'teleport', 'x': tx, 'y': ty})

# Wait 10 seconds, checking if position changes
print("  等待10秒观察位置...", flush=True)
for i in range(10):
    time.sleep(1)
    if last_x == tx and last_y == ty:
        print(f"  [!!!] 瞬移成功! 位置=({last_x},{last_y})", flush=True)
        break
else:
    print(f"  [-] 位置未变为目标: 当前=({last_x},{last_y}) 目标=({tx},{ty})", flush=True)

script.post({'type': 'stop_tp'})
time.sleep(1)

# Test 2: Try smaller teleport (+20)
tx2 = (last_x + 20) & 0xFF
ty2 = (last_y + 20) & 0xFF
print(f"\n[TEST2] 小步瞬移: ({last_x},{last_y}) -> ({tx2},{ty2})", flush=True)
script.post({'type': 'teleport', 'x': tx2, 'y': ty2})
for i in range(8):
    time.sleep(1)
    if last_x == tx2 and last_y == ty2:
        print(f"  [!!!] 瞬移成功! 位置=({last_x},{last_y})", flush=True)
        break
else:
    print(f"  [-] 位置未变为目标: 当前=({last_x},{last_y})", flush=True)

script.post({'type': 'stop_tp'})
time.sleep(1)

# Test 3: Speed hack 3x
print(f"\n[TEST3] 3x速度测试...", flush=True)
script.post({'type': 'speed', 'mult': 3.0})
print("  移动角色观察是否加速... (10秒)", flush=True)
time.sleep(10)
script.post({'type': 'speed', 'mult': 1.0})

print("\n[*] 测试完成", flush=True)
time.sleep(2)
session.detach()
print("Done.", flush=True)

File: tools/teleport_rapid.js
var GAME_FD = %d;
var libc = Process.getModuleByName("libc.so");
var M1 = 0, M2 = 0;
var lastX = null, lastY = null;
var targetX = null, targetY = null;
var posAddr = null;
var writeInterval = null;
var speedMult = 1.0;

// Phase 1: find position address using 12-byte pattern
var scanPhase = 0;
var prevMatches = [];

function collectRanges() {
    var all = [];
    var ranges = Process.enumerateRanges({protection: 'rw-', coalesce: true});
    ranges.forEach(function(r) {
        var sz = typeof r.size === 'number' ? r.size : Number(r.size);
        if (sz >= 65536 && sz <= 67108864) {
            all.push({base: r.base, size: sz});
        }
    });
    return all;
}

function scanXY12(ranges, x, y) {
    var pat = ('0' + x.toString(16)).slice(-2) + ' 00 00 00 ?? ?? ?? ?? ' +
              ('0' + y.toString(16)).slice(-2) + ' 00 00 00';
    var res = [];
    ranges.forEach(function(r) {
        try {
            var m = Memory.scanSync(r.base, r.size, pat);
            for (var i = 0; i < m.length; i++) {
                res.push(m[i].address.toString());
                if (res.length >= 100) return;
            }
        } catch(e) {}
    });
    return res;
}

// Rapid write to memory
function rapidWrite() {
    if (!posAddr || targetX === null) return;
    try {
        ptr(posAddr).writeU32(targetX);
        ptr(posAddr).add(8).writeU32(targetY);
    } catch(e) {}
}

Interceptor.attach(libc.getExportByName("send"), {
    onEnter: function(args) {
        var fd = args[0].toInt32();
        var buf = args[1];
        var len = args[2].toInt32();
        if (fd === GAME_FD && len === 30 && buf.readU8() === 3) {
            var key = buf.add(1).readU8();
            var p = [];
            for (var i = 0; i < 29; i++) p.push(buf.add(i + 1).readU8() ^ key);
            var x = p[17], y = p[21];
            if (!M1) { M1 = p[1]; M2 = p[3]; }

            // Phase 0-1: find position address (same as before)
            if (!posAddr && lastX !== null && (x !== lastX || y !== lastY)) {
                if (scanPhase === 0) {
                    var ranges = collectRanges();
                    prevMatches = scanXY12(ranges, x, y);
                    send({t: 'scan1', x: x, y: y, count: prevMatches.length});
                    scanPhase = 1;
                } else if (scanPhase === 1) {
                    var ranges = collectRanges();
                    var newMatches = scanXY12(ranges, x, y);
                    send({t: 'scan2', x: x, y: y, count: newMatches.length});

                    var prevSet = {};
                    for (var i = 0; i < prevMatches.length; i++) prevSet[prevMatches[i]] = true;

                    for (var i = 0; i < newMatches.length; i++) {
                        if (prevSet[newMatches[i]]) {
                            posAddr = newMatches[i];
                            send({t: 'found', addr: posAddr, x: x, y: y});
                            break;
                        }
                    }
                    if (!posAddr) {
                        prevMatches = newMatches;
                        send({t: 'retry'});
                    } else {
                        scanPhase = 2;
                    }
                }
            }

            // Modify packet position if target is set
            if (targetX !== null && targetY !== null) {
                buf.add(18).writeU8(targetX ^ key);
                buf.add(20).writeU8(targetX ^ M1 ^ key);
                buf.add(22).writeU8(targetY ^ key);
                buf.add(24).writeU8(targetY ^ M2 ^ key);
            }

            // Speed hack: amplify position delta in packet
            if (speedMult !== 1.0 && lastX !== null && targetX === null) {
                var dx = x - lastX;
                var dy = y - lastY;
                var nx = (lastX + Math.round(dx * speedMult)) & 0xFF;
                var ny = (lastY + Math.round(dy * speedMult)) & 0xFF;
                buf.add(18).writeU8(nx ^ key);
                buf.add(20).writeU8(nx ^ M1 ^ key);
                buf.add(22).writeU8(ny ^ key);
                buf.add(24).writeU8(ny ^ M2 ^ key);
            }

            if (lastX === null || x !== lastX || y !== lastY) {
                // Also read memory for verification
                var memInfo = '';
                if (posAddr) {
                    try {
                        var mx = ptr(posAddr).readU32() & 0xFF;
                        var my = ptr(posAddr).add(8).readU32() & 0xFF;
                        if (mx !== x || my !== y) {
                            memInfo = ' mem=(' + mx + ',' + my + ')';
                        }
                    } catch(e) {}
                }
                send({t: 'pos', x: x, y: y, mem: memInfo, target: targetX !== null});
            }
            lastX = x; lastY = y;
        }
    }
});

// Commands
recv('teleport', function(data) {
    targetX = data.x & 0xFF;
    targetY = data.y & 0xFF;
    // Start rapid write interval (every 30ms)
    if (writeInterval) clearInterval(writeInterval);
    if (posAddr) {
        writeInterval = setInterval(rapidWrite, 30);
        send({t: 'tp_start', x: targetX, y: targetY, addr: posAddr});
    } else {
        send({t: 'tp_noaddr', x: targetX, y: targetY});
    }
});

recv('stop_tp', function(data) {
    targetX = null; targetY = null;
    if (writeInterval) { clearInterval(writeInterval); writeInterval = null; }
    send({t: 'tp_stop'});
});

recv('speed', function(data) {
    speedMult = data.mult;
    send({t: 'speed_set', mult: speedMult});
});

send({t: 'ready'});
